File: apps/worker/internal/db/dal/tags.go
package dal

import (
	"context"

	"github.com/uptrace/bun"

	"github.com/kunish/wheel/apps/worker/internal/types"
)

func ListTags(ctx context.Context, db *bun.DB) ([]types.Tag, error) {
	var tags []types.Tag
	err := db.NewSelect().Model(&tags).OrderExpr("id ASC").Scan(ctx)
	if err != nil {
		return nil, err
	}
	if tags == nil {
		tags = []types.Tag{}
	}
	return tags, nil
}

func CreateTag(ctx context.Context, db *bun.DB, tag *types.Tag) error {
	_, err := db.NewInsert().Model(tag).Exec(ctx)
	return err
}

func UpdateTag(ctx context.Context, db *bun.DB, id int, data map[string]any) error {
	allowed := map[string]bool{
		"name": true, "color": true, "description": true,
	}
	q := db.NewUpdate().Table("tags")
	count := 0
	for col, val := range data {
		if allowed[col] {
			q = q.Set(col+" = ?", val)
			count++
		}
	}
	if count == 0 {
		return nil // nothing to update
	}
	_, err := q.Where("id = ?", id).Exec(ctx)
	return err
}

func DeleteTag(ctx context.Context, db *bun.DB, id int) error {
	_, err := db.NewDelete().Model((*types.Tag)(nil)).Where("id = ?", id).Exec(ctx)
	return err
}

File: apps/web/src/lib/api/tags.ts
import { apiFetch } from "./client"

// ── Tags ──

export interface Tag {
  id: number
  name: string
  color: string
  description: string
  channelCount: number
  keyCount: number
  createdAt?: string
}

export interface TagInput {
  id?: number
  name: string
  color: string
  description: string
}

export function listTags() {
  return apiFetch<{ success: boolean; data: { tags: Tag[] } }>("/api/v1/tag/list")
}

export function createTag(data: Omit<TagInput, "id">) {
  return apiFetch<{ success: boolean; data: Tag }>("/api/v1/tag/create", {
    method: "POST",
    body: data,
  })
}

export function updateTag(data: Partial<TagInput> & { id: number }) {
  return apiFetch<{ success: boolean }>("/api/v1/tag/update", {
    method: "POST",
    body: data,
  })
}

export function deleteTag(id: number) {
  return apiFetch<{ success: boolean }>(`/api/v1/tag/delete/${id}`, {
    method: "DELETE",
  })
}
